Lib/test/test_modulefinder.py

```diff
@@ -299,6 +291,16 @@ def _do_test(self, info, report=False, debug=0, replace_paths=[]):
             bad, maybe = mf.any_missing_maybe()
             self.assertEqual(bad, missing)
             self.assertEqual(maybe, maybe_missing)
+
+            # check for modules encoding
+            if encoding:
+                for module in mf.modules.items():
+                    module_name, module_object = module
+                    file = open(module_object.__file__, 'rb')
+                    encoding = tokenize.detect_encoding(file.readline)[0]
+                    self.assertEqual(encoding, 'iso-8859-1')
+                    file.close()
```

Comment:
just a tip for python, you generally don't want to manually open / close files like this -- usually prefer the `with` statement (to avoid leaking file descriptors in the exceptional case):

```python
with open(...) as ...:
    ...
```

(this will automatically close the file for you!)

Reply:
Cool, thanks!

Reply:
I've updated the test with your suggestion but it's not working. Could you take a look please?

Reply:
```diff
diff --git a/Lib/test/test_modulefinder.py b/Lib/test/test_modulefinder.py
index ebd96e1c8a..376fa1ea76 100644
--- a/Lib/test/test_modulefinder.py
+++ b/Lib/test/test_modulefinder.py
@@ -342,6 +342,12 @@ class ModuleFinderTest(unittest.TestCase):
         os.remove(source_path)
         self._do_test(bytecode_test)
 
+    def test_non_utf8_encoding(self):
+        os.makedirs(TEST_DIR, exist_ok=True)
+        with open(os.path.join(TEST_DIR, 'a.py'), 'w', encoding='cp1252') as f:
+            f.write('# -*- coding: cp1252 -*-\nx = "€"\n')
+        self._do_test(["a", ["a"], [], [], ""])
+
     def test_replace_paths(self):
         old_path = os.path.join(TEST_DIR, 'a', 'module.py')
         new_path = os.path.join(TEST_DIR, 'a', 'spam.py')
```

here ya go!

before your patch it fails like this:

```console
$ ./python  -m test.test_modulefinder
.....E........
======================================================================
ERROR: test_non_utf8_encoding (__main__.ModuleFinderTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/asottile/workspace/cpython/Lib/test/test_modulefinder.py", line 349, in test_non_utf8_encoding
    self._do_test(["a", ["a"], [], [], ""])
  File "/home/asottile/workspace/cpython/Lib/test/test_modulefinder.py", line 281, in _do_test
    mf.import_hook(import_this)
  File "/home/asottile/workspace/cpython/Lib/modulefinder.py", line 177, in import_hook
    q, tail = self.find_head_package(parent, name)
  File "/home/asottile/workspace/cpython/Lib/modulefinder.py", line 233, in find_head_package
    q = self.import_module(head, qname, parent)
  File "/home/asottile/workspace/cpython/Lib/modulefinder.py", line 326, in import_module
    m = self.load_module(fqname, fp, pathname, stuff)
  File "/home/asottile/workspace/cpython/Lib/modulefinder.py", line 343, in load_module
    co = compile(fp.read()+'\n', pathname, 'exec')
  File "/home/asottile/workspace/cpython/Lib/codecs.py", line 322, in decode
    (result, consumed) = self._buffer_decode(data, self.errors, final)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 30: invalid start byte

----------------------------------------------------------------------
Ran 14 tests in 0.127s

FAILED (errors=1)
```

after your patch it succeeds:

```console
$ ./python  -m test.test_modulefinder
..............
----------------------------------------------------------------------
Ran 14 tests in 0.115s

OK
```

Reply:
Cool! Thank you so much!